packages/input-text/src/TextInput.js

```diff
@@ -0,0 +1,165 @@
+      invalid: this.state.invalid,
+      disabled: otherProps.disabled
+    };
+    const inputProps = {
+      value: this.state.value,
+      onChange: this.onChange,
+      onInvalid: this.onInvalid,
+      onMouseOver: this.onMouseOver,
+      onMouseLeave: this.onMouseLeave,
+      ...otherProps
+    };
+    return (
+      <StyledLabel {...containerProps}>
+        {this._renderInput(inputProps)}
+        {this._renderLabel(label)}
+        {this._renderTooltip(this.state.showTooltip, tooltipText)}
+        {this._renderFailedMessages(
+          this.state.invalid,
+          validationErrors,
+          errorIconSrc
+        )}
+      </StyledLabel>
+    );
+  }
+
+  _renderLabel(label) {
+    return label && <div className="label">{label}</div>;
+  }
+
+  _renderInput(inputProps) {
+    return <input type="text" {...inputProps} />;
+  }
+
+  _renderTooltip(visible, text) {
+    return (
+      visible && (
+        <div className="tooltip">
+          <Tooltip text={text} />
+        </div>
+      )
+    );
+  }
+
+  _renderFailedMessages(invalid, errors, iconSrc) {
+    return (
+      invalid && (
+        <div className="messages">
+          {errors.map((text, index) => (
+            <div className="message" key={`err-${index}`}>
+              <Image
+                className="icon"
+                src={iconSrc}
+                badge
+                size={badgeSizes.SMALL}
+              />
+              {text}
+            </div>
+          ))}
+        </div>
+      )
+    );
+  }
+  onChange(event) {
+    this.setState({
+      value: event.target.value,
+      invalid: false
+    });
+    return this.props.onChange(event);
+  }
+
+  onInvalid(event) {
+    event.preventDefault();
+    this.setState({ invalid: true });
+    return this.props.onInvalid(event);
+  }
+
+  onMouseOver(event) {
+    if (this.props.disabled) {
+      this.setState({
+        showTooltip: true
+      });
+    }
+    return this.props.onMouseOver(event);
+  }
+
+  onMouseLeave(event) {
+    if (this.props.disabled) {
+      this.setState({
+        showTooltip: false
+      });
+    }
+    return this.props.onMouseLeave(event);
+  }
+
+  static propTypes = {
+    label: PropTypes.string,
+    value: PropTypes.string,
+    onChange: PropTypes.func,
+    invalid: PropTypes.bool,
+    validationErrors: PropTypes.arrayOf(PropTypes.string),
+    errorIconSrc: PropTypes.string,
+    tooltipText: PropTypes.string,
+    onInvalid: PropTypes.func,
+    onMouseOver: PropTypes.func,
+    onMouseLeave: PropTypes.func,
+
+    placeholder: PropTypes.string,
+    disabled: PropTypes.bool,
+    required: PropTypes.bool
+    // and any other properties to forward to the html input element...
+  };
+
+  static defaultProps = {
+    value: "",
+    invalid: false,
+    onChange: () => null,
+    onInvalid: () => null,
+    onMouseOver: () => null,
+    onMouseLeave: () => null,
+    errorIconSrc,
+    validationErrors: ["This field is required"],
```

Comment:
I don't think we should have a `default` validation message... 
What if we controll the `invalid` state with this field? I mean: if we have `validationErrors.length` > 0, field is invalid.

Reply:
ok, let me rethink this api then... 

having no default messages, would make the native bubble that I am preventing default behavior less useful, and I think it would be nice to mirror html's form validation interface. I will try to purpose something better, and if I fail I can implement your proposed implicit api 
(no messages array = no validation erros)